### PR TITLE
Fix incorrect interpolation of elevation

### PR DESCRIPTION
**Pull request by rotu**
Previously there was a mistake in the elevation calculation. The weights for interpolating a height were swapped.

This is likely only visible if you're using 3D rendering, are zoomed in close, and you're doing something that reacts to small changes in elevation (like updating a map marker to remain under the mouse cursor).

https://github.com/maplibre/maplibre-gl-js/assets/119948/ad8f1e8d-d5d1-4c0e-8401-560c6f622578

https://github.com/maplibre/maplibre-gl-js/assets/119948/6ed0a3c0-126b-49c7-965e-f78be090f110

## Launch Checklist

 - [X] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [X] Briefly describe the changes in this PR.
 - [X] Link to related issues.
 - [x] Include before/after visuals or gifs if this PR includes visual changes.
 - [x] Write tests for all new functionality.
 - [ ] Document any changes to public APIs.
 - [ ] Post benchmark scores.
 - [ ] Add an entry to `CHANGELOG.md` under the `## main` section.

### CODE REVIEW

**Comment by rotu** [comment]
I suspect the render tests need to be updated but they fail with uninformative error messages for me when run locally:

`1/1238: failed tests/zoomed-raster/underzoom undefined`

**Comment by HarelM** [comment]
The CI is saying that there are differences between the images, not sure why you get this results locally.
Also the CI report is uploaded and you can download the new image and replace it to fix the tests.

**Comment by rotu** [comment]
> The CI is saying that there are differences between the images, not sure why you get this results locally. Also the CI report is uploaded and you can download the new image and replace it to fix the tests.

Yeah, I knew there were difference between images. I think the render test runner might be broken on Mac. It's certainly not ideal to have tests that can't run locally.

Thanks - I didn't see the build artifacts and didn't even think about pulling the images from there!

**Comment by prozessor13** [comment]
Wow, thank you very much for finding this error!! :D 

Although your fix works, my original error was in the tl (=topleft), tr (=topright), bl (=bottomleft), br (=bottomright) calculation, not the interpolation, where you made the fix. Details in my code-comment.

**Comment by rotu** [comment]
I don’t see the code-comment of which you speak.

Merely looking at the code, it’s not clear which direction is “up” and what direction is “left” so I think that naming convention might do more harm than good.

File: src/render/terrain.ts [hunk 1]
```diff
@@ -138,7 +138,7 @@ export class Terrain {
             const tr = terrain.tile.dem.get(c[0], c[1] + 1);
             const bl = terrain.tile.dem.get(c[0] + 1, c[1]);
             const br = terrain.tile.dem.get(c[0] + 1, c[1] + 1);
-            elevation = interpolates.number(interpolates.number(tl, tr, coord[0] - c[0]), interpolates.number(bl, br, coord[0] - c[0]), coord[1] - c[1]);
+            elevation = interpolates.number(interpolates.number(tl, tr, coord[1] - c[1]), interpolates.number(bl, br, coord[1] - c[1]), coord[0] - c[0]);
```
Comment (prozessor13) [inline src/render/terrain.ts:141]: I think the correct calculation for tl, tr, bl, br should be:
````
            const tl = terrain.tile.dem.get(c[0], c[1]);
            const tr = terrain.tile.dem.get(c[0] + 1, c[1]);
            const bl = terrain.tile.dem.get(c[0], c[1] + 1);
            const br = terrain.tile.dem.get(c[0] + 1, c[1] + 1);
````
then, i think, the interplation can be as it was, and all variables are named correctly.

**Comment by prozessor13** [comment]
> I don’t see the code-comment of which you speak.

Do you see my comment now?

**Comment by prozessor13** [comment]
> Merely looking at the code, it’s not clear which direction is “up” and what direction is “left” so I think that naming convention might do more harm than good.

ok, please feel free to rename the variables.

File: src/render/terrain.ts (hunk 1 above)
Comment (rotu) [reply src/render/terrain.ts:141]: What do "top", "bottom" "left" and "right" mean?  The `x` and `y` arguments are simply numbers and there's no indication of the coordinate system in the contextual code. I feel like the code is missing a comment or documentation somewhere.

File: src/render/terrain.ts [hunk 2]
```diff
@@ -128,19 +127,25 @@ export class Terrain {
...
+        const cx = Math.floor(coord[0]),
+            cy = Math.floor(coord[1]),
+            tx = coord[0] - cx,
+            ty = coord[1] - cy;
+        return (
+            dem.get(cx, cy) * (1 - tx) * (1 - ty) +
+            dem.get(cx + 1, cy) * (tx) * (1 - ty) +
+            dem.get(cx, cy + 1) * (1 - tx) * (ty) +
+            dem.get(cx + 1, cy + 1) * (tx) * (ty)
+        );
```
Comment (prozessor13) [inline src/render/terrain.ts:148]: Great! This code looks very good!

**Comment by HarelM** [comment]
Can you please add tests to the code that was added?

**Comment by rotu** [comment]
> Can you please add tests to the code that was added?

Tests added. The code is structured pretty poorly for testing and unfortunately has to assume quite a few implementation details. I feel dirty.

File: src/render/terrain.test.ts [hunk 3]
```diff
@@ -162,4 +163,37 @@ describe('Terrain', () => {
...
     });
+
+    test('interpolation works', () => {
+        // const mockTerrain = jest.spyOn(Terrain)
```
Comment (HarelM) [inline src/render/terrain.test.ts:168]: Can these comments be removed?
Although I wish it was possible to use these, the getTerrainData method is hard to pass by mocking other stuff... I understand you dirtiness feeling...

Comment (rotu) [reply src/render/terrain.test.ts]: Yep. I removed those comments and forgot to push :-).

**Review by HarelM — APPROVED** [review]

**Pull request merged**